### PR TITLE
feat: handle genesis

### PR DESCRIPTION
**Pull request by n0izn0iz**
WIP

### CODE REVIEW

**Comment by jefft0** [comment]
Related to https://github.com/gnolang/gno/pull/1941

**Comment by zivkovicmilos** [comment]
@n0izn0iz 
I think you can just add a special "genesis" fetch before we start the main run routines for fetching blocks / block result chunks, since for the genesis block we only need the block results (block 0 is unfetchable) as part of https://github.com/gnolang/gno/pull/1941

**Comment by n0izn0iz** [comment]
agree, would simplify, was planning on doing that

**Review by ajnavarro — CHANGES_REQUESTED** [review]
This is a breaking change, existing indexers won't get genesis data unless they delete the storage.

Could you add some tests please?

File: fetch/fetch.go [hunk 1]
```diff
@@ -67,6 +70,46 @@ func New(
 	return f
 }
 
+func (f *Fetcher) FetchGenesisData() error {
+	_, err := f.storage.GetLatestHeight()
+	isInit := errors.Is(err, storageErrors.ErrNotFound)
+
+	if !isInit {
+		return nil
+	}
```
[the remark this answers is not in the document]
Comment (n0izn0iz) [reply fetch/fetch.go]: done in 40cb71105d9be5c17a315dcd35d2483cb2d300a7

File: fetch/fetch.go [hunk 2]
```diff
@@ -67,6 +70,46 @@ func New(
...
+		}
+
+		txResults[txIndex] = result
+	}
+
+	if err := f.writeBatch([]*bft_types.Block{block}, [][]*bft_types.TxResult{txResults}, 0, 1); err != nil {
+		return err
+	}
+
+	return nil
```
[the remark this answers is not in the document]
Comment (n0izn0iz) [reply fetch/fetch.go]: done in b9b7e16a421b61054d33feed68bbc983c9b90189

File: fetch/fetch.go [hunk 3]
```diff
@@ -178,68 +221,115 @@ func (f *Fetcher) FetchChainData(ctx context.Context) error {
...
+func getGenesisBlock(client Client) (*bft_types.Block, error) {
+	gblock, err := client.GetGenesis()
+	if err != nil {
+		return nil, fmt.Errorf("unable to get genesis block, %w", err)
+	}
+
+	genesisState, ok := gblock.Genesis.AppState.(gnoland.GnoGenesisState)
+	if !ok {
+		return nil, fmt.Errorf("unknown genesis state kind")
+	}
```
[the remark this answers is not in the document]
Comment (n0izn0iz) [reply fetch/fetch.go]: done in 24e01699983df1a22e8573fba2e3b1038c4abc57

**Comment by n0izn0iz** [comment]
added tests in 763c68ac29fa7f90b43673205e0d9ba82bb6039f

**Review by ajnavarro — CHANGES_REQUESTED** [review]
Please, check the tests. Mocks are used to simulate dependencies our system interacts with. This helps us focus on testing the real functionality without checking the mock itself.

To make sure our tests are testing what is needed, we need to test not just the happy path but also different possible errors. Here are some scenarios we should include:

- Remote Server Failure: What happens if the remote server fails? We need to make sure our system handles server issues properly.
- Genesis Retrieval Issues: What happens if the genesis data can’t be retrieved? This could be due to network problems, data corruption, or other issues.
- Unexpected Genesis Object Type: What happens if the genesis object is not the expected type? Type mismatches can cause unexpected behavior, so our tests should check how the system deals with this.

These scenarios are important, but we should also think of other possible cases to ensure our system is robust and reliable.

File: fetch/fetch_test.go [hunk 4]
```diff
@@ -903,6 +904,125 @@ func TestFetcher_InvalidBlocks(t *testing.T) {
...
+		require.NoError(t, err)
+
+		expected := &types.TxResult{
+			Height:   0,
+			Index:    i,
+			Tx:       amino.MustMarshalJSON(txs[i]),
+			Response: abci.ResponseDeliverTx{},
+		}
+		require.Equal(t, expected, tx)
+	}
```
[the remark this answers is not in the document]
Comment (n0izn0iz) [reply fetch/fetch_test.go]: we are testing that the storage did get properly populated but I see your point of view. I removed the `GetBlock`/`GetTx` implems and checked the local variables directly instead

**Comment by n0izn0iz** [comment]
improved tests in [0a31d61](https://github.com/gnolang/tx-indexer/pull/34/commits/0a31d61dbc390dbc1b055304d4fe50a996748e55)

it helped me catch nil deref corner cases, thanks

File: fetch/fetch.go [hunk 5]
```diff
@@ -178,68 +219,121 @@ func (f *Fetcher) FetchChainData(ctx context.Context) error {
 				// Pop the next chunk
 				f.chunkBuffer.PopFront()
 
-				wb := f.storage.WriteBatch()
+				if err := f.writeBatch(
+					item.chunk.blocks,
+					item.chunk.results,
+					item.chunkRange.from,
```
Comment (zivkovicmilos) [inline fetch/fetch.go:225]: The `from` and `to` args are used only for the logger, you can move them out of this method and log after

File: fetch/fetch.go [hunk 6]
```diff
@@ -67,6 +70,44 @@ func New(
...
+			Height:   0,
+			Index:    uint32(txIndex),
+			Tx:       tx,
+			Response: results.Results.DeliverTxs[txIndex],
+		}
+
+		txResults[txIndex] = result
+	}
+
+	return f.writeBatch([]*bft_types.Block{block}, [][]*bft_types.TxResult{txResults}, 0, 1)
```
Comment (zivkovicmilos) [inline fetch/fetch.go:108]: Wouldn't the range here be `0, 0`, since we're writing a single block?

**Review by zivkovicmilos — CHANGES_REQUESTED** [review]
Looking at the implementation now, I think we made a right choice in the relevant gno genesis PR 👍 

I've left a few comments that should be addressed before we merge, but nothing groundbreaking 🙏

File: fetch/fetch.go [hunk 7]
```diff
@@ -67,6 +70,44 @@ func New(
 	return f
 }
 
+func (f *Fetcher) FetchGenesisData() error {
+	_, err := f.storage.GetLatestHeight()
```
[the remark this answers is not in the document]
Comment (n0izn0iz) [reply fetch/fetch.go:74]: Let's look at the cases:

- `err` is `nil`: the storage has a latest height, this means at least the genesis data has been written, or some blocks past it, we do nothing and return `nil`
- `err` is `ErrNotFound`: the storage is empty, we execute the rest of the routine and fetch genesis data
- `err` is something else: there has been a storage error, we do nothing and return this error

I agree it's not super readable, do you want me to add a comment?

File: fetch/mocks_test.go [hunk 8]
```diff
@@ -41,6 +45,14 @@ func (m *mockClient) GetBlock(blockNum uint64) (*core_types.ResultBlock, error)
 	return nil, nil
 }
 
+func (m *mockClient) GetGenesis() (*core_types.ResultGenesis, error) {
+	if m.getGenesisFn != nil {
+		return m.getGenesisFn()
+	}
+
+	return &core_types.ResultGenesis{Genesis: &bft_types.GenesisDoc{AppState: gnoland.GnoGenesisState{}}}, nil
```
[the remark this answers is not in the document]
Comment (n0izn0iz) [reply fetch/mocks_test.go]: done in 0632c4384fffc4e9841a290d68637de89b6d35a1

File: fetch/fetch.go [hunk 9]
```diff
@@ -178,68 +219,121 @@ func (f *Fetcher) FetchChainData(ctx context.Context) error {
...
+	gblock, err := client.GetGenesis()
+	if err != nil {
+		return nil, fmt.Errorf("unable to get genesis block: %w", err)
+	}
+
+	if gblock.Genesis == nil {
+		return nil, fmt.Errorf("nil genesis doc")
+	}
+
+	appState := gblock.Genesis.AppState
```
[the remark this answers is not in the document]
Comment (n0izn0iz) [reply fetch/fetch.go]: done in 4e17d199b9b3f7dec77860696351a526d694ca7e

File: fetch/fetch.go [hunk 10]
```diff
@@ -178,68 +219,121 @@ func (f *Fetcher) FetchChainData(ctx context.Context) error {
...
+
+	genesisState, ok := appState.(gnoland.GnoGenesisState)
+	if !ok {
```
[the remark this answers is not in the document]
Comment (n0izn0iz) [reply fetch/fetch.go:312]: the `Client` is an interface, there could be an implementation that returns another type

File: fetch/fetch.go [hunk 11]
```diff
@@ -178,68 +219,121 @@ func (f *Fetcher) FetchChainData(ctx context.Context) error {
...
-					return fmt.Errorf("error persisting block information into storage, %w", err)
-				}
-			}
+func getGenesisBlock(client Client) (*bft_types.Block, error) {
+	gblock, err := client.GetGenesis()
+	if err != nil {
+		return nil, fmt.Errorf("unable to get genesis block: %w", err)
+	}
+
+	if gblock.Genesis == nil {
```
[the remark this answers is not in the document]
Comment (n0izn0iz) [reply fetch/fetch.go:307]: the `Client` is an interface, there could be a bad implementation that returns a malformed type, I think it's better to fail gracefully instead of crashing with nil deref

File: cmd/start.go [hunk 12]
```diff
@@ -159,6 +159,10 @@ func (c *startCfg) exec(ctx context.Context) error {
 		fetch.WithMaxChunkSize(c.maxChunkSize),
 	)
 
+	if err := f.FetchGenesisData(); err != nil {
```
[the remark this answers is not in the document]
Comment (n0izn0iz) [reply cmd/start.go:162]: putting FetchGenesisData in FetchChainData will require me to change almost all tests, I tried and it was really hard to understand what was the right thing to do

I was thinking about renaming FetchChainData to something else (fetchRemainingData or something) and create a new FetchChainData that just calls fetchGenesisData and fetchRemainingData so I can leave the tests as is

File: fetch/fetch.go (hunk 5 above)
Comment (n0izn0iz) [reply fetch/fetch.go]: I changed the method to accept a slot directly instead of it's fields separately

File: fetch/fetch.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (zivkovicmilos) [reply fetch/fetch.go:74]: > I agree it's not super readable, do you want me to add a comment?

Yes please 🙏

File: cmd/start.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (zivkovicmilos) [reply cmd/start.go:162]: >putting FetchGenesisData in FetchChainData will require me to change almost all tests, I tried and it was really hard to understand what was the right thing to do

This is the option I think we should go with -- what tests were failing?

File: fetch/fetch.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (n0izn0iz) [reply fetch/fetch.go:74]: done in [42b9dc8](https://github.com/gnolang/tx-indexer/pull/34/commits/42b9dc8c781b63ee49f402d6b65ddabf869515fe)

File: cmd/start.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (n0izn0iz) [reply cmd/start.go]: I moved it then ([583a722](https://github.com/gnolang/tx-indexer/pull/34/commits/583a722e6b5cff93f2be4d539479e4ab4445c7e6)), help is very welcome

**Comment by n0izn0iz** [comment]
there is one remaining problem, the indexer should ignore genesis result if the node returns the well known "height must be greater than 0" otherwise we can't index test4 anymore

**Comment by zivkovicmilos** [comment]
@n0izn0iz I'll fix up the tests on this branch 🙏

**Comment by zivkovicmilos** [comment]
@n0izn0iz 

I think we got the PR to a working state -- but there is a problem:

The indexer will break for networks that don't have the genesis results in their storage.

I've circumvented this problem by providing the genesis state fetch error as a soft one, which doesn't derail the indexer:
[ca459d3](https://github.com/gnolang/tx-indexer/pull/34/commits/ca459d371990a62ba6ce35b727c612b56b06380a)

**Review by zivkovicmilos — APPROVED** [review]
Thank you for the fixup 🙏

**Review by ajnavarro — APPROVED** [review]
Tested locally, it won't break current indexes.
